src/golang/cmd/server/handler/connect_integration.go

```diff
@@ -202,58 +207,241 @@ func (h *ConnectIntegrationHandler) Perform(ctx context.Context, interfaceArgs i
+			// While we don't enforce this atomically, we can make the two update together to minimize the risk.
+			err = h.updateStorageMigrationExecState(ctx, storageMigrationObj.ID, &execState)
+			if err != nil {
+				log.Errorf("Unexpected error when updating the storage migration entry to SUCCESS: %v", err)
+				return
+			}
+
+			err = config.UpdateStorage(storageConfig)
+			if err != nil {
+				log.Errorf("Unexpected error when updating the global storage layer config: %v", err)
+			} else {
+				log.Info("Successfully updated the global storage layer config!")
+			}
 		}()
 	}
 
 	return emptyResp, http.StatusOK, nil
 }
 
-// ConnectIntegration connects a new integration specified by `args`. It returns a status code for the request
-// and an error, if any.
+// Also updates `current=True` if the execution state is marked as SUCCESS!
+func (h *ConnectIntegrationHandler) updateStorageMigrationExecState(
+	ctx context.Context,
+	storageMigrationID uuid.UUID,
+	execState *shared.ExecutionState,
+) error {
+	updates := map[string]interface{}{
+		models.StorageMigrationExecutionState: execState,
+	}
+
+	// This is updated to a transaction if we also need to mark an old entry as current=False.
+	db := h.Database
+	if execState.Status == shared.SucceededExecutionStatus {
+		updates[models.StorageMigrationCurrent] = true
+
+		// If there was a previous storage migration, update that entry to be `current=False`.
+		oldStorageMigrationObj, err := h.StorageMigrationRepo.Current(ctx, h.Database)
+		if err == nil {
+			// Updating the old storage migration entry must be done in the same transaction
+			// as the update of the new storage migration entry, so that there is at most one
+			// current=True entry in the storage_migration table.
+			txn, err := h.Database.BeginTx(context.Background())
+			if err != nil {
+				return errors.Wrap(err, "Unable to start transaction for updating storage state.")
+			}
+			defer database.TxnRollbackIgnoreErr(ctx, txn)
+
+			db = txn
+			_, err = h.StorageMigrationRepo.Update(
+				ctx,
+				oldStorageMigrationObj.ID,
+				map[string]interface{}{
+					models.StorageMigrationCurrent: false,
+				},
+				db,
+			)
+			if err != nil {
+				return errors.Wrap(err, "Unexpected error when updating old storage migration entry to be non-current")
+			}
+		} else if !aq_errors.Is(err, database.ErrNoRows()) {
+			return errors.Wrap(err, "Unexpected error when fetching current storage state.")
+		}
+		// Continue without doing anything if there was no previous storage migration.
+	}
+
+	// Perform the actual intended execution state update.
+	_, err := h.StorageMigrationRepo.Update(
+		ctx,
+		storageMigrationID,
+		updates,
+		db,
+	)
+	if err != nil {
+		return errors.Wrap(err, "Unexpected error when updating storage migration execution state.")
+	}
+
+	// Only need to do this if we're committing a transaction.
+	if txn, ok := db.(database.Transaction); ok {
```

Comment:
What does this block do? It feels like an antipattern to me - all transaction owner (whoever calls txn := db.BeginTx()) should be responsible for calling `Commit()` and `TxnRollBack()`. I'm confused why we need to access some transaction this helper didn't create itself.

Reply:
It's a little confusing, I agree. Refactored it so that the transaction definition happens at the top of the function.